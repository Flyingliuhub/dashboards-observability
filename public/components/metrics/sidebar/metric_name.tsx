/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React, { useEffect, useState } from 'react';
import { EuiFacetButton, EuiIcon } from '@elastic/eui';

interface IMetricNameProps {
  metric: any;
  handleClick: (props: any) => void;
  showDeleteIcon: boolean;
}

export const MetricName = (props: IMetricNameProps) => {
  const { metric, handleClick, showDeleteIcon } = props;

  const icons = {
    empty: <EuiIcon type="empty" />,
    danger: <EuiIcon type="trash" color="danger" />,
    primary: <EuiIcon type="plusInCircleFilled" color="primary" />,
  };

  const [iconDisplay, setIconDisplay] = useState(false);
  const [icon, setIcon] = useState(icons.empty);

  const setIconFromButtonState = (buttonState: string) => {
    return setIcon(icons[buttonState]);
  };

  useEffect(() => {
    console.log({ iconDisplay, icon });
  }, [iconDisplay, icon]);

  useEffect(() => {
    const iconState = !iconDisplay ? 'empty' : showDeleteIcon ? 'danger' : 'primary';
    setIconFromButtonState(iconState);
  }, [iconDisplay]);

  const name = (metricName: string) => {
    return metric.catalog === 'CUSTOM_METRICS' ? metricName : metricName.split('.')[1];
  };

  return (
    <EuiFacetButton
      icon={icon}
      quantity={2}
      onClick={() => handleClick(metric)}
      onMouseOver={() => setIconDisplay(true)}
      onMouseOut={() => setIconDisplay(false)}
    >
      {metric.name}
    </EuiFacetButton>
  );
};
