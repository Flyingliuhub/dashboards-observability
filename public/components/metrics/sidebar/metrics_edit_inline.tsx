/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React, { useEffect, useState } from 'react';
import {
  EuiComboBox,
  EuiExpression,
  EuiFlexGroup,
  EuiFlexItem,
  EuiFormLabel,
  EuiPopover,
  EuiPopoverTitle,
  EuiSelect,
} from '@elastic/eui';
import { from } from 'rxjs';
import { distinct, filter, map, mergeMap, tap, toArray } from 'rxjs/operators';
import { MetricType } from '../../../../common/types/metrics';
import { AGGREGATION_OPTIONS } from '../../../../common/constants/metrics';

const useObservable = (observable, observer, deps) => {
  // useEffect with empty deps will call this only once
  useEffect(() => {
    const sub = observable.subscribe(observer); // connect
    return () => sub.unsubscribe(); // < unsub on unmount
  }, deps);
};

export const MetricsEditInline = ({
  visualizationData,
  metricMetaData,
  updateMetricConfig,
}: {
  visualizationData?: any;
  metricMetaData?: MetricType;
  updateMetricConfig: ({}: { aggregation?: string; attributesGroupBy?: string[] }) => void;
}) => {
  const [aggregationIsOpen, setAggregationIsOpen] = useState(false);
  const [attributesGroupByIsOpen, setAttributesGroupByIsOpen] = useState(false);
  const [availableAttributes, setAvailableAttributes] = useState([]);

  useObservable(
    from(visualizationData?.datarows || []).pipe(
      mergeMap((row) => Object.keys(row[0])),
      filter((attributeKey) => attributeKey !== '__name__'),
      distinct((attributeKey) => attributeKey),
      map((attributeKey) => ({ label: attributeKey, value: attributeKey })),
      toArray()
    ),
    setAvailableAttributes,
    [visualizationData]
  );

  // useEffect(() => {
  //   console.log({ availableAttributes });
  // }, [availableAttributes]);

  const onChangeAggregation = (e) => {
    updateMetricConfig({ aggregation: e.target.value });
    setAggregationIsOpen(false);
  };

  const onChangeAttributesGroupBy = (e) => {
    updateMetricConfig({ selectedAttributesGroupBy: e.target.value });
    setAttributesGroupByIsOpen(false);
  };

  useEffect(() => {
    console.log({ availableAttributes, metricMetaData, visualizationData });
  }, [availableAttributes, metricMetaData, visualizationData]);

  const renderAggregationEditor = () => (
    <div>
      <EuiPopoverTitle>Aggregation</EuiPopoverTitle>
      <EuiSelect
        compressed
        value={metricMetaData.query.aggregation}
        onChange={onChangeAggregation}
        options={AGGREGATION_OPTIONS}
      />
    </div>
  );

  const renderAttributesGroupByEditor = () => (
    <div>
      <EuiPopoverTitle>Group By Attributes</EuiPopoverTitle>
      <EuiComboBox
        compressed
        value={metricMetaData.query.attributesGroupBy}
        onChange={onChangeAggregation}
        options={[{ label: 'host' }]}
      />
    </div>
  );

  return (
    <div>
      <EuiFlexGroup>
        <EuiFlexItem grow={false}>
          <EuiPopover
            id="aggregation"
            button={
              <EuiExpression
                description="aggregation"
                value={metricMetaData.query.aggregation}
                isActive={aggregationIsOpen}
                onClick={() => setAggregationIsOpen(true)}
              />
            }
            isOpen={aggregationIsOpen}
            closePopover={() => setAggregationIsOpen(false)}
            panelPaddingSize="s"
            anchorPosition="downLeft"
          >
            {renderAggregationEditor()}
          </EuiPopover>
        </EuiFlexItem>
        <EuiFlexItem grow={false}>
          <EuiPopover
            id="attributes"
            button={
              <EuiExpression
                description="group by attributes"
                value={'(none selected)'}
                isActive={attributesGroupByIsOpen}
                onClick={() => setAttributesGroupByIsOpen(true)}
              />
            }
            isOpen={attributesGroupByIsOpen}
            closePopover={() => setAttributesGroupByIsOpen(false)}
            panelPaddingSize="s"
            anchorPosition="downLeft"
          >
            {renderAttributesGroupByEditor()}
          </EuiPopover>
        </EuiFlexItem>
      </EuiFlexGroup>
    </div>
  );
};
