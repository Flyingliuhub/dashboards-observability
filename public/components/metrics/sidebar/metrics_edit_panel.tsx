/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import {
  EuiComboBox,
  EuiFlexGroup,
  EuiFlexItem,
  EuiForm,
  EuiFormRow,
  EuiSelect,
} from '@elastic/eui';
import React, { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { AGGREGATION_OPTIONS } from '../../../../common/constants/metrics';
import { metricQuerySelector, updateMetricQuery } from '../redux/slices/metrics_slice';

export const MetricsEditPanel = ({ visualizationId }: { visualizationId: string }) => {
  // const [selectedAttributes, setSelectedAttributes] = useState('');
  // const [selectedAggregation, setSelectedAggregation] = useState('avg');

  const dispatch = useDispatch();

  const query = useSelector(metricQuerySelector(visualizationId));
  useEffect(() => {
    console.log({ visualizationId, query });
  }, [query, visualizationId]);

  const availableAttributesLabels = query.availableAttributes.map((attribute) => ({
    label: attribute,
    name: attribute,
  }));

  const onChangeAggregation = (e) => {
    dispatch(updateMetricQuery(visualizationId, { aggregation: e.target.value }));
  };

  const onChangeAttributesGroupBy = async (selectedAttributes) => {
    console.log('onChangeAttributes', selectedAttributes);
    const attributesGroupBy = selectedAttributes.map(({ label }) => label);
    dispatch(updateMetricQuery(visualizationId, { attributesGroupBy }));
  };

  const flexForm = (
    <EuiFlexGroup direction="column">
      <EuiFlexItem grow={false}>
        <EuiSelect
          compressed
          value={query.aggregation}
          onChange={onChangeAggregation}
          options={AGGREGATION_OPTIONS}
        />
      </EuiFlexItem>
      <EuiFlexItem grow={false}>
        <EuiComboBox
          placeholder="Select grouping attributes"
          options={availableAttributesLabels}
          selectedOptions={query.attributesGroupBy.map((label) => ({ label, value: label }))}
          onChange={onChangeAttributesGroupBy}
          isClearable={true}
          data-test-subj="metrics__attributesComboBox"
        />
      </EuiFlexItem>
    </EuiFlexGroup>
  );
  const formGroup = (
    <EuiForm>
      <EuiFormRow label="Aggregation">
        <EuiSelect
          compressed
          value={query.aggregation}
          onChange={onChangeAggregation}
          options={AGGREGATION_OPTIONS}
        />
      </EuiFormRow>
      <EuiFormRow label="Aggregations Group By">
        <EuiComboBox
          placeholder="Select grouping attributes"
          options={availableAttributesLabels}
          selectedOptions={query.attributesGroupBy.map((label) => ({ label, value: label }))}
          onChange={onChangeAttributesGroupBy}
          isClearable={true}
          data-test-subj="metrics__attributesComboBox"
        />
      </EuiFormRow>
    </EuiForm>
  );

  return formGroup;
};
