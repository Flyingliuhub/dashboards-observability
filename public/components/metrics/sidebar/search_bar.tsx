/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { EuiSearchBar } from '@elastic/eui';
import React from 'react';
import { useDispatch } from 'react-redux';
import { debounce } from 'lodash';
import { clearSearchedMetrics, searchMetric } from '../redux/slices/metrics_slice';

interface ISearchBarProps {
  setSearch: React.Dispatch<React.SetStateAction<boolean>>;
}

export const SearchBar = (props: ISearchBarProps) => {
  const dispatch = useDispatch();

  const onChange = debounce(({ query }: { query: any }) => {
    if (query.text !== '') {
      dispatch(searchMetric(query.text));
    } else {
      dispatch(clearSearchedMetrics());
    }
  }, 300);

  return (
    <div className="metrics-search-bar-input" data-test-subj="metricsSearch">
      <EuiSearchBar
        box={{
          placeholder: 'Search for metrics',
          incremental: true,
        }}
        defaultQuery={''}
        onChange={onChange}
      />
    </div>
  );
};
