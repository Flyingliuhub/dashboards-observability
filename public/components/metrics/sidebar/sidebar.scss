/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

.metricsList {
  line-height: normal;
}

.metricsListContainer {
  margin-left: 15px;
  margin-right: 5px;
}

.tokenMargin {
  vertical-align: middle;
}

#metrics__mainContent {
.dscAppContainer{
.metricsSideBar {
  padding: $euiSizeS;
      overflow: auto;

      &__fieldGroups {
        @include euiYScrollWithShadows;

        overflow-y: auto;
        margin-right: -$euiSizeS;
        padding-right: $euiSizeS;
        margin-top: $euiSizeS;
      }

      &__fieldGroup {
        margin-top: $euiSizeS;

        &:first-child {
          margin-top: 0;
        }
      }
    }
    .metricsSideBar {
  @include euiYScrollWithShadows;
  .sidebar-list {
    height: 100%;
  }
}}}
