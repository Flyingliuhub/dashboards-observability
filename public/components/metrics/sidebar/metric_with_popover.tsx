/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React, { useState } from 'react';

import {
  EuiPopover,
  EuiButton,
  EuiText,
  EuiForm,
  EuiFormRow,
  EuiSelect,
  EuiComboBox,
} from '@elastic/eui';
import { MetricName } from './metric_name';

export const MetricWithPopover = ({ metric }) => {
  const [isPopoverOpen, setIsPopoverOpen] = useState(false);

  const onButtonClick = () => setIsPopoverOpen((currOpen) => !currOpen);
  const closePopover = () => setIsPopoverOpen(false);

  const attributeOptions = [
    {
      label: 'host',
    },

    {
      label: 'path',
    },
    {
      label: 'method',
    },
  ];

  return (
    <EuiPopover
      button={<MetricName metric={metric} handleClick={onButtonClick} />}
      isOpen={isPopoverOpen}
      closePopover={closePopover}
      anchorPosition="rightUp"
    />
  );
};
