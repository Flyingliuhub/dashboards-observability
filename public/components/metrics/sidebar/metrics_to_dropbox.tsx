/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import { MetricName } from './metric_name';

export const mapMetricsToSelectedPanel = (metrics) => {
  const panelComponents = metrics.map((metric) => {
    return <MetricName metric={metric} />;
  });

  return (
    <>
      <div className="vbConfig__content">{panelComponents}</div>
    </>
  );
};
