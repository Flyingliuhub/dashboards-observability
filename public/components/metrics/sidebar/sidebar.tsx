/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import './sidebar.scss';

import React, { useEffect } from 'react';
import { EuiSpacer } from '@elastic/eui';
import { I18nProvider } from '@osd/i18n/react';
import {
  useMetricStore,
  availableMetricsSelector,
  selectedMetricsSelector,
  loadMetrics,
} from '../metrics_store';
import { MetricsAccordion } from './metrics_accordion';
import { SearchBar } from './search_bar';

export const Sidebar = () => {
  const availableMetrics = useMetricStore(availableMetricsSelector);
  const selectedMetrics = useMetricStore(selectedMetricsSelector);
  const { deSelectMetric, selectMetric } = useMetricStore();

  useEffect(() => {
    loadMetrics();
  }, []);

  return (
    <I18nProvider>
      <section className="sidebarHeight">
        <SearchBar />
        <EuiSpacer size="s" />

        <MetricsAccordion
          metricsList={selectedMetrics}
          headerName="Selected Metrics"
          handleClick={deSelectMetric}
          dataTestSubj="metricsListItems_selectedMetrics"
        />
        <EuiSpacer size="s" />
        <MetricsAccordion
          metricsList={availableMetrics}
          headerName="Available Metrics"
          handleClick={selectMetric}
          dataTestSubj="metricsListItems_availableMetrics"
        />
      </section>
    </I18nProvider>
  );
};
