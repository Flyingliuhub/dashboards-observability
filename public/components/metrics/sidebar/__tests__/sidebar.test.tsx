/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';
import { waitFor } from '@testing-library/react';
import httpClientMock from '../../../../../test/__mocks__/httpClientMock';
import PPLService from '../../../../services/requests/ppl';
import { applyMiddleware, createStore } from '@reduxjs/toolkit';
import { rootReducer } from '../../../../framework/redux/reducers';
import { Provider } from 'react-redux';
import { Sidebar } from '../sidebar';
import thunk from 'redux-thunk';
import { coreRefs } from '../../../../framework/core_refs';
import { sampleSavedMetric } from '../../../../../test/metrics_contants';
import { SavedObjectsActions } from '../../../../services/saved_objects/saved_object_client/saved_objects_actions';
import { act } from 'react-dom/test-utils';
import ReactDOM from 'react-dom';

const flushPromises = () => new Promise(process.nextTick);

describe('Side Bar Component', () => {
  configure({ adapter: new Adapter() });

  it('renders Side Bar Component for Saved Metric', async () => {
    jest.useFakeTimers();

    console.log('*** CUSTOM ***');

    SavedObjectsActions.getBulk = jest
      .fn()
      .mockResolvedValue({ observabilityObjectList: [sampleSavedMetric] });

    httpClientMock.get = jest.fn();

    coreRefs.pplService = new PPLService(httpClientMock);
    coreRefs.pplService.fetch = jest.fn().mockResolvedValueOnce({
      data: { DATASOURCE_NAME: [] },
    });

    const el = document.createElement('div');
    let wrapper;
    await act(async () => {
      jest.advanceTimersByTime(1);
      wrapper = mount(<Sidebar />);
      await Promise.resolve();
      wrapper.update();
      // ReactDOM.render(<Sidebar />, el);
    });

    expect(wrapper).toContain('Available Metrics 0 of 0');

    await act(async () => {
      jest.advanceTimersByTime(1);
      await Promise.resolve();
      wrapper.update();
    });

    expect(wrapper).toContain('Available Metrics 1 of 1');
    expect(wrapper).toMatchSnapshot();

    jest.useRealTimers();
  });

  it('renders Side Bar Component for Prometheus Metric', async () => {
    jest.clearAllMocks();
    jest.clearAllTimers();
    jest.useFakeTimers();
    console.log('*** PROMETHEUS ***');
    SavedObjectsActions.getBulk = jest.fn().mockResolvedValue({ observabilityObjectList: [] });

    httpClientMock.get = jest.fn();

    coreRefs.pplService = new PPLService(httpClientMock);
    coreRefs.pplService.fetch = jest
      .fn()
      .mockResolvedValueOnce({
        data: { DATASOURCE_NAME: ['datasource1'] },
      })
      .mockResolvedValueOnce({
        jsonData: [
          {
            TABLE_CATALOG: 'datasource1',
            TABLE_NAME: 'metric1',
            TABLE_TYPE: 'metric',
          },
        ],
      });

    const el = document.createElement('div');
    let wrapper;
    await act(async () => {
      jest.advanceTimersByTime(1);
      ReactDOM.render(<Sidebar />, el);
      jest.runAllTicks();
      await Promise.resolve();
    });

    expect(el.innerHTML).toContain('Available Metrics 0 of 0');

    await act(async () => {
      jest.advanceTimersByTime(1);
      jest.runAllTicks();
      await Promise.resolve();
    });

    expect(el.innerHTML).toContain('Available Metrics 1 of 1');
    expect(el).toMatchSnapshot();

    jest.useRealTimers();
  });
});
