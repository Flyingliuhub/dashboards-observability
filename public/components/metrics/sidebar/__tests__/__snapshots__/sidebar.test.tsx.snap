// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Side Bar Component renders Side Bar Component for Prometheus Metric 1`] = `
<div>
  <section
    class="sidebarHeight"
  >
    <div
      class="metrics-search-bar-input"
      data-test-subj="metricsSearch"
    >
      <div
        class="euiFlexGroup euiFlexGroup--gutterMedium euiFlexGroup--alignItemsCenter euiFlexGroup--directionRow euiFlexGroup--responsive euiFlexGroup--wrap"
      >
        <div
          class="euiFlexItem euiSearchBar__searchHolder"
        >
          <div
            class="euiFormControlLayout euiFormControlLayout--fullWidth"
          >
            <div
              class="euiFormControlLayout__childrenWrapper"
            >
              <input
                aria-label="This is a search bar. As you type, the results lower in the page will automatically filter."
                class="euiFieldSearch euiFieldSearch--fullWidth"
                placeholder="Search for metrics"
                type="search"
                value=""
              />
              <div
                class="euiFormControlLayoutIcons"
              >
                <span
                  class="euiFormControlLayoutCustomIcon"
                >
                  <svg
                    aria-hidden="true"
                    class="euiIcon euiIcon--medium euiIcon-isLoading euiFormControlLayoutCustomIcon__icon"
                    focusable="false"
                    height="16"
                    role="img"
                    viewBox="0 0 16 16"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  />
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="euiSpacer euiSpacer--s"
    />
    <div
      class="euiAccordion euiAccordion-isOpen"
    >
      <div
        class="euiAccordion__triggerWrapper"
      >
        <button
          aria-controls="Selected MetricsSelector"
          aria-expanded="true"
          class="euiAccordion__button"
          id="random_html_id"
          type="button"
        >
          <span
            class="euiAccordion__iconWrapper"
          >
            <svg
              aria-hidden="true"
              class="euiIcon euiIcon--medium euiIcon-isLoading euiAccordion__icon euiAccordion__icon-isOpen"
              focusable="false"
              height="16"
              role="img"
              viewBox="0 0 16 16"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            />
          </span>
          <span
            class="euiIEFlexWrapFix"
          >
            <span
              class="euiTitle euiTitle--xxxsmall"
            >
              Selected Metrics
               
              0
               of 
              0
            </span>
          </span>
        </button>
      </div>
      <div
        aria-labelledby="random_html_id"
        class="euiAccordion__childWrapper"
        id="Selected MetricsSelector"
        role="region"
        style="height: 0px"
        tabindex="-1"
      >
        <div>
          <div
            class=""
          >
            <ul
              class="metricsList"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="euiSpacer euiSpacer--s"
    />
    <div
      class="euiAccordion euiAccordion-isOpen"
    >
      <div
        class="euiAccordion__triggerWrapper"
      >
        <button
          aria-controls="Available MetricsSelector"
          aria-expanded="true"
          class="euiAccordion__button"
          id="random_html_id"
          type="button"
        >
          <span
            class="euiAccordion__iconWrapper"
          >
            <svg
              aria-hidden="true"
              class="euiIcon euiIcon--medium euiIcon-isLoading euiAccordion__icon euiAccordion__icon-isOpen"
              focusable="false"
              height="16"
              role="img"
              viewBox="0 0 16 16"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            />
          </span>
          <span
            class="euiIEFlexWrapFix"
          >
            <span
              class="euiTitle euiTitle--xxxsmall"
            >
              Available Metrics
               
              1
               of 
              1
            </span>
          </span>
        </button>
      </div>
      <div
        aria-labelledby="random_html_id"
        class="euiAccordion__childWrapper"
        id="Available MetricsSelector"
        role="region"
        style="height: 0px"
        tabindex="-1"
      >
        <div>
          <div
            class=""
          >
            <ul
              class="metricsList"
            >
              <li
                class="metricsListContainer"
                data-test-subj="metricsListItems_availableMetrics"
              >
                <button
                  class="euiFacetButton euiFacetButton--unSelected obsMetric-Name eui-textTruncate"
                  title="datasource1.metric1"
                  type="button"
                >
                  <span
                    class="euiFacetButton__content"
                  >
                    <div
                      aria-label="datasource1"
                      class="euiAvatar euiAvatar--s euiAvatar--space"
                      style="background-color: rgb(121, 170, 217); color: rgb(0, 0, 0);"
                      title="datasource1"
                    >
                      <span
                        aria-hidden="true"
                      >
                        d
                      </span>
                    </div>
                    <span
                      class="euiFacetButton__text"
                      data-text="metric1"
                    >
                      metric1
                    </span>
                  </span>
                </button>
              </li>
            </ul>
          </div>
        </div>
      </div>
    </div>
  </section>
</div>
`;

exports[`Side Bar Component renders Side Bar Component for Saved Metric 1`] = `
<div>
  <section
    class="sidebarHeight"
  >
    <div
      class="metrics-search-bar-input"
      data-test-subj="metricsSearch"
    >
      <div
        class="euiFlexGroup euiFlexGroup--gutterMedium euiFlexGroup--alignItemsCenter euiFlexGroup--directionRow euiFlexGroup--responsive euiFlexGroup--wrap"
      >
        <div
          class="euiFlexItem euiSearchBar__searchHolder"
        >
          <div
            class="euiFormControlLayout euiFormControlLayout--fullWidth"
          >
            <div
              class="euiFormControlLayout__childrenWrapper"
            >
              <input
                aria-label="This is a search bar. As you type, the results lower in the page will automatically filter."
                class="euiFieldSearch euiFieldSearch--fullWidth"
                placeholder="Search for metrics"
                type="search"
                value=""
              />
              <div
                class="euiFormControlLayoutIcons"
              >
                <span
                  class="euiFormControlLayoutCustomIcon"
                >
                  <svg
                    aria-hidden="true"
                    class="euiIcon euiIcon--medium euiIcon-isLoading euiFormControlLayoutCustomIcon__icon"
                    focusable="false"
                    height="16"
                    role="img"
                    viewBox="0 0 16 16"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  />
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="euiSpacer euiSpacer--s"
    />
    <div
      class="euiAccordion euiAccordion-isOpen"
    >
      <div
        class="euiAccordion__triggerWrapper"
      >
        <button
          aria-controls="Selected MetricsSelector"
          aria-expanded="true"
          class="euiAccordion__button"
          id="random_html_id"
          type="button"
        >
          <span
            class="euiAccordion__iconWrapper"
          >
            <svg
              aria-hidden="true"
              class="euiIcon euiIcon--medium euiIcon-isLoading euiAccordion__icon euiAccordion__icon-isOpen"
              focusable="false"
              height="16"
              role="img"
              viewBox="0 0 16 16"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            />
          </span>
          <span
            class="euiIEFlexWrapFix"
          >
            <span
              class="euiTitle euiTitle--xxxsmall"
            >
              Selected Metrics
               
              0
               of 
              0
            </span>
          </span>
        </button>
      </div>
      <div
        aria-labelledby="random_html_id"
        class="euiAccordion__childWrapper"
        id="Selected MetricsSelector"
        role="region"
        tabindex="-1"
      >
        <div>
          <div
            class=""
          >
            <ul
              class="metricsList"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="euiSpacer euiSpacer--s"
    />
    <div
      class="euiAccordion euiAccordion-isOpen"
    >
      <div
        class="euiAccordion__triggerWrapper"
      >
        <button
          aria-controls="Available MetricsSelector"
          aria-expanded="true"
          class="euiAccordion__button"
          id="random_html_id"
          type="button"
        >
          <span
            class="euiAccordion__iconWrapper"
          >
            <svg
              aria-hidden="true"
              class="euiIcon euiIcon--medium euiIcon-isLoading euiAccordion__icon euiAccordion__icon-isOpen"
              focusable="false"
              height="16"
              role="img"
              viewBox="0 0 16 16"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            />
          </span>
          <span
            class="euiIEFlexWrapFix"
          >
            <span
              class="euiTitle euiTitle--xxxsmall"
            >
              Available Metrics
               
              1
               of 
              1
            </span>
          </span>
        </button>
      </div>
      <div
        aria-labelledby="random_html_id"
        class="euiAccordion__childWrapper"
        id="Available MetricsSelector"
        role="region"
        tabindex="-1"
      >
        <div>
          <div
            class=""
          >
            <ul
              class="metricsList"
            >
              <li
                class="metricsListContainer"
                data-test-subj="metricsListItems_availableMetrics"
              >
                <button
                  class="euiFacetButton euiFacetButton--unSelected obsMetric-Name eui-textTruncate"
                  title="[Logs] Average ram usage per day by windows os "
                  type="button"
                >
                  <span
                    class="euiFacetButton__content"
                  >
                    <svg
                      aria-hidden="true"
                      aria-labelledby="random_html_id"
                      class="euiIcon euiIcon--large euiIcon-isLoading"
                      focusable="false"
                      height="16"
                      role="img"
                      viewBox="0 0 16 16"
                      width="16"
                      xmlns="http://www.w3.org/2000/svg"
                    />
                    <span
                      class="euiFacetButton__text"
                      data-text="[Logs] Average ram usage per day by windows os "
                    >
                      [Logs] Average ram usage per day by windows os 
                    </span>
                  </span>
                </button>
              </li>
            </ul>
          </div>
        </div>
      </div>
    </div>
  </section>
</div>
`;

exports[`renders Side Bar Component for Saved Metric 1`] = `
<div>
  <section
    class="sidebarHeight"
  >
    <div
      class="metrics-search-bar-input"
      data-test-subj="metricsSearch"
    >
      <div
        class="euiFlexGroup euiFlexGroup--gutterMedium euiFlexGroup--alignItemsCenter euiFlexGroup--directionRow euiFlexGroup--responsive euiFlexGroup--wrap"
      >
        <div
          class="euiFlexItem euiSearchBar__searchHolder"
        >
          <div
            class="euiFormControlLayout euiFormControlLayout--fullWidth"
          >
            <div
              class="euiFormControlLayout__childrenWrapper"
            >
              <input
                aria-label="This is a search bar. As you type, the results lower in the page will automatically filter."
                class="euiFieldSearch euiFieldSearch--fullWidth"
                placeholder="Search for metrics"
                type="search"
                value=""
              />
              <div
                class="euiFormControlLayoutIcons"
              >
                <span
                  class="euiFormControlLayoutCustomIcon"
                >
                  <svg
                    aria-hidden="true"
                    class="euiIcon euiIcon--medium euiIcon-isLoading euiFormControlLayoutCustomIcon__icon"
                    focusable="false"
                    height="16"
                    role="img"
                    viewBox="0 0 16 16"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  />
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="euiSpacer euiSpacer--s"
    />
    <div
      class="euiAccordion euiAccordion-isOpen"
    >
      <div
        class="euiAccordion__triggerWrapper"
      >
        <button
          aria-controls="Selected MetricsSelector"
          aria-expanded="true"
          class="euiAccordion__button"
          id="random_html_id"
          type="button"
        >
          <span
            class="euiAccordion__iconWrapper"
          >
            <svg
              aria-hidden="true"
              class="euiIcon euiIcon--medium euiIcon-isLoading euiAccordion__icon euiAccordion__icon-isOpen"
              focusable="false"
              height="16"
              role="img"
              viewBox="0 0 16 16"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            />
          </span>
          <span
            class="euiIEFlexWrapFix"
          >
            <span
              class="euiTitle euiTitle--xxxsmall"
            >
              Selected Metrics
               
              0
               of 
              0
            </span>
          </span>
        </button>
      </div>
      <div
        aria-labelledby="random_html_id"
        class="euiAccordion__childWrapper"
        id="Selected MetricsSelector"
        role="region"
        tabindex="-1"
      >
        <div>
          <div
            class=""
          >
            <ul
              class="metricsList"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="euiSpacer euiSpacer--s"
    />
    <div
      class="euiAccordion euiAccordion-isOpen"
    >
      <div
        class="euiAccordion__triggerWrapper"
      >
        <button
          aria-controls="Available MetricsSelector"
          aria-expanded="true"
          class="euiAccordion__button"
          id="random_html_id"
          type="button"
        >
          <span
            class="euiAccordion__iconWrapper"
          >
            <svg
              aria-hidden="true"
              class="euiIcon euiIcon--medium euiIcon-isLoading euiAccordion__icon euiAccordion__icon-isOpen"
              focusable="false"
              height="16"
              role="img"
              viewBox="0 0 16 16"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            />
          </span>
          <span
            class="euiIEFlexWrapFix"
          >
            <span
              class="euiTitle euiTitle--xxxsmall"
            >
              Available Metrics
               
              1
               of 
              1
            </span>
          </span>
        </button>
      </div>
      <div
        aria-labelledby="random_html_id"
        class="euiAccordion__childWrapper"
        id="Available MetricsSelector"
        role="region"
        tabindex="-1"
      >
        <div>
          <div
            class=""
          >
            <ul
              class="metricsList"
            >
              <li
                class="metricsListContainer"
                data-test-subj="metricsListItems_availableMetrics"
              >
                <button
                  class="euiFacetButton euiFacetButton--unSelected obsMetric-Name eui-textTruncate"
                  title="[Logs] Average ram usage per day by windows os "
                  type="button"
                >
                  <span
                    class="euiFacetButton__content"
                  >
                    <svg
                      aria-hidden="true"
                      aria-labelledby="random_html_id"
                      class="euiIcon euiIcon--large euiIcon-isLoading"
                      focusable="false"
                      height="16"
                      role="img"
                      viewBox="0 0 16 16"
                      width="16"
                      xmlns="http://www.w3.org/2000/svg"
                    />
                    <span
                      class="euiFacetButton__text"
                      data-text="[Logs] Average ram usage per day by windows os "
                    >
                      [Logs] Average ram usage per day by windows os 
                    </span>
                  </span>
                </button>
              </li>
            </ul>
          </div>
        </div>
      </div>
    </div>
  </section>
</div>
`;
